torchgeo/datasets/utils.py

```diff
@@ -631,6 +635,74 @@ def path_is_vsi(path: Path) -> bool:
     return '://' in str(path) or str(path).startswith('/vsi')
 
 
+def listdir_vsi_recursive(root: Path) -> list[str]:
+    """Lists all files in Virtual File Systems (VSI) recursively.
+
+    Args:
+        root: directory to list. These must contain the prefix for the VSI
+            e.g. '/vsiaz/' or 'az://' for azure blob storage
+                 '/vsizip/' or 'zip://' for zipped archives
+
+    Returns:
+        A list of all file paths matching filename_glob in the root VSI directory or its
+        subdirectories.
+
+    .. versionadded:: #0.6
+    """
+    dirs = [str(root)]
+    files = []
+    while dirs:
+        dir = dirs.pop()
+        try:
+            subdirs = fiona.listdir(dir)
+            dirs.extend([os.path.join(dir, subdir) for subdir in subdirs])
+        except FionaValueError as e:
+            if 'is not a directory' in str(e):
+                # Assuming dir is a file as it is not a directory
+                # fiona.listdir can throw FionaValueError for only two reasons
+                files.append(dir)
+            else:
+                # fiona.listdir can throw FionaValueError for only two reasons
+                # 1. 'is not a directory'
+                # 2. 'does not exist'
+                raise
+    return files
+
+
+def list_directory_recursive(root: Path, filename_glob: str) -> list[str]:
```

Comment:
Should we return a set instead of a list? The `files` method is the only place they are used and uses a set.

Reply:
Thats ok by me. The only reason I did not is because similar methods returns a list (iglob, fnmatch etc).

Reply:
We could change this method to remove duplicates (set) and sort it, then return a list as well. Or is it better to be explicit about that in the files-property?